main.js

```diff
@@ -16,6 +17,121 @@ function changeGroup(str) {
+  await exec.exec(`powershell.exe`, [`(Get-FileHash ${destination} -Algorithm SHA256)[0].Hash`], {listeners: {stdout: (data) => { computedChecksum += data.toString(); }}});
+  if (computedChecksum.slice(0, -2).toUpperCase() !== checksum.toUpperCase()) {
+    throw new Error(`The SHA256 of the installer does not match! expected ${checksum} got ${computedChecksum}`);
+  }
+}
+
+async function disableKeyRefresh(msysRootDir) {
+  const postFile = path.join(msysRootDir, 'etc\\post-install\\07-pacman-key.post');
+  await exec.exec(`powershell.exe`, [`((Get-Content -path ${postFile} -Raw) -replace '--refresh-keys', '--version') | Set-Content -Path ${postFile}`]);
+}
+
+
+class PackageCache {
+
+  constructor(msysRootDir, input) {
+    // We include "update" in the fallback key so that a job run with update=false never fetches
+    // a cache created with update=true. Because this would mean a newer version than needed is in the cache
+    // which would never be used but also would win during cache prunging because it is newer.
+    this.fallbackCacheKey = 'msys2-pkgs-upd:' + input.update.toString();
+
+    // We want a cache key that is ideally always the same for the same kind of job.
+    // So that mingw32 and ming64 jobs, and jobs with different install packages have different caches.
+    let shasum = crypto.createHash('sha1');
+    shasum.update([input.release, input.update, input.pathtype, input.msystem, input.install].toString());
+    this.jobCacheKey = this.fallbackCacheKey + '-conf:' + shasum.digest('hex').slice(0, 8);
+
+    this.restoreKey = undefined;
+    this.pkgCachePath = path.join(msysRootDir, 'var', 'cache', 'pacman', 'pkg');
+  }
+
+  async restore() {
+    // We ideally want a cache matching our configuration, but every cache is OK since we prune it later anyway
+    this.restoreKey = await cache.restoreCache([this.pkgCachePath], this.jobCacheKey, [this.jobCacheKey, this.fallbackCacheKey]);
+    console.log(`Cache restore for ${this.jobCacheKey}, got ${this.restoreKey}`);
+  }
+
+  async save() {
+    const saveKey = this.jobCacheKey + '-files:' + (await hashElement(this.pkgCachePath))['hash'].toString();
+    if (this.restoreKey === saveKey) {
+      console.log(`Cache unchanged, skipping save for ${saveKey}`);
+    } else {
+      try {
+        const cacheId  = await cache.saveCache([this.pkgCachePath], saveKey);
+        console.log(`Cache saved as ID ${cacheId} using key ${saveKey}`);
+      } catch (error) {
+        // In case something created the same cache since we restored we'll get an error here,
+        // but that's OK with us, so ignore.
+        console.log(error.message);
+      }
+    }
+  }
+
+  async prune() {
+    // Remove all uninstalled packages
+    await runMsys(['paccache', '-r', '-f', '-u', '-k0']);
+    // Keep the newest for all other packages
+    await runMsys(['paccache', '-r', '-f', '-k1']);
+  }
+}
+
+let cmd = null;
+
+async function writeWrapper(msysRootDir, input, destDir, name) {
```

Comment:
It seems that `input.pathtype` is used only. I'd change the argument to `pathtype`.